### PR TITLE
[WIP] Support value encoding

### PR DESCRIPTION
**Pull request by divarvel**

### CODE REVIEW

File: src/encode.rs [hunk 1]
```diff
@@ -9,10 +9,14 @@
...
 use std::io;
 use std::io::Write;
+use std::mem;
+use Value;
+use Schema;
 
 #[derive(Debug)]
 pub enum EncodeErrorKind {
     WriteFailed(io::Error),
+    InvalidSchema
```
Comment (jminer) [inline src/encode.rs]: Nit: Could you end this line with a comma?

File: src/encode.rs [hunk 2]
```diff
@@ -79,3 +83,134 @@ fn test_encode_zig_zag() {
     assert_eq!(encode_zig_zag(i64::min_value()), 0xFFFFFFFF_FFFFFFFF);
     assert_eq!(encode_zig_zag(i64::max_value()), 0xFFFFFFFF_FFFFFFFE); // 0x7FFF...FF
 }
+
+pub fn encode<W: Write>(writer: &mut W, schema: &Schema, value: &Value) -> Result<(), EncodeError> {
+    match (schema, value) {
+        (&Schema::Null, &Value::Null) => Ok(()),
+        (&Schema::Null, _)            => Err(EncodeError { kind: EncodeErrorKind::InvalidSchema }),
+
+        (&Schema::Boolean, &Value::Boolean(true))  => { writer.write(&[0x01]); Ok(()) },
```
Comment (jminer) [inline src/encode.rs]: Could you wrap `try!()` around the `write()` call here and in the rest of the function?

**Comment by jminer** [comment]
Sorry for the delay (October is busy for me). It looks good. I'd mainly like to see `try!()` added to propagate errors.

**Comment by divarvel** [comment]
I'm very busy as well :-S I'll do the changes in a few days when I'm available

**Comment by divarvel** [comment]
I've wrapped the `writer.write` and the recursive `encode` calls with `try!`

**Comment by jminer** [comment]
Looking at the Avro C API, I noticed that it doesn't pass the schema when encoding anything but a record. If `encode`'s signature were

```
fn encode<W: Write>(writer: &mut W, value: &Value)
```

Then `encode` could still check the schema of each field of a `Value::Record` (because it has a reference to the schema), and other types would just be encoded. I don't know if that's better or not.

Thanks for the PR! Feel free to open more if you get a chance.

**Pull request merged**